File: android/src/org/tmurakam/cashflow/models/Ledger.java
// -*-  Mode:java; c-basic-offset:4; tab-width:4; indent-tabs-mode:t -*-

// Ledger : 総勘定元帳

package org.tmurakam.cashflow.models;

import java.util.*;

import android.database.sqlite.*;

import org.tmurakam.cashflow.ormapper.*;

public class Ledger {
	public ArrayList<Asset> assets = null;

	public void load() {
		assets = Asset.find_cond("ORDER BY sorder");
	}

	public void rebuild() {
		for (Asset as : assets) {
			as.rebuild();
		}
	}

	public int assetCount() {
		return assets.size();
	}

	public Asset assetAtIndex(int n) {
		return assets.get(n);
	}

	public Asset assetWithKey(int pid) {
		for (Asset as : assets) {
			if (as.pid == pid) return as;
		}
		return null;
	}


	public int assetIndexWithKey(int pid) {
		int i;
		for (i = 0; i < assets.size(); i++) {
			Asset as = assets.get(i);
			if (as.pid == pid) return i;
		}
		return -1;	  
	}

	public void addAsset(Asset as) {
		assets.add(as);
		as.insert();
	}

	public void deleteAsset(Asset as) {
		as.delete();

		DataModel.getJournal().deleteAllTransactionsWithAsset(as);

		assets.remove(as);
		rebuild();
	}

	public void updateAsset(Asset asset) {
		asset.update();
	}

	public void reorderAsset(int from, int to) {
		Asset as = assets.get(from);
		assets.remove(from);
		assets.add(to, as);
	
		// renumbering sorder
		SQLiteDatabase db = Database.instance();
		db.beginTransaction();
		for (int i = 0; i < assets.size(); i++) {
			as = assets.get(i);
			as.sorder = i;
			as.update();
		}
		db.endTransaction();
	}
}

File: Classes/models/Ledger.m
// -*-  Mode:ObjC; c-basic-offset:4; tab-width:8; indent-tabs-mode:nil -*-
/*
 * CashFlow for iOS
 * Copyright (C) 2008-2011, Takuya Murakami, All rights reserved.
 * For conditions of distribution and use, see LICENSE file.
 */

// Ledger : 総勘定元帳

#import "DataModel.h"
#import "Ledger.h"

@implementation Ledger

@synthesize assets = mAssets;

- (void)load
{
    self.assets = [Asset find_all:@"ORDER BY sorder"];
}

- (void)rebuild
{
    for (Asset *as in mAssets) {
        [as rebuild];
    }
}

- (int)assetCount
{
    return [mAssets count];
}

- (Asset*)assetAtIndex:(int)n
{
    return [mAssets objectAtIndex:n];
}

- (Asset*)assetWithKey:(int)pid
{
    for (Asset *as in mAssets) {
        if (as.pid == pid) return as;
    }
    return nil;
}

- (int)assetIndexWithKey:(int)pid
{
    int i;
    for (i = 0; i < [mAssets count]; i++) {
        Asset *as = [mAssets objectAtIndex:i];
        if (as.pid == pid) return i;
    }
    return -1;
}

- (void)addAsset:(Asset *)as
{
    [mAssets addObject:as];
    [as save];
}

- (void)deleteAsset:(Asset *)as
{
    [as delete];

    [[DataModel journal] deleteAllTransactionsWithAsset:as];

    [mAssets removeObject:as];

    [self rebuild];
}

- (void)updateAsset:(Asset*)asset
{
    [asset save];
}

- (void)reorderAsset:(int)from to:(int)to
{
    Asset *as = [[mAssets objectAtIndex:from] retain];
    [mAssets removeObjectAtIndex:from];
    [mAssets insertObject:as atIndex:to];
    [as release];
	
    // renumbering sorder
    Database *db = [Database instance];
    [db beginTransaction];
    for (int i = 0; i < [mAssets count]; i++) {
        as = [mAssets objectAtIndex:i];
        as.sorder = i;
        [as save];
    }
    [db commitTransaction];
}

@end
